src/strategies/gadget.rs

```diff
@@ -0,0 +1,261 @@
+use crate::{mds_matrix::MDS_MATRIX, Curve, Scalar, WIDTH};
+
+use super::Strategy;
+
+use num_traits::{One, Zero};
+use plonk::cs::{composer::StandardComposer, constraint_system::Variable};
+
+/// Implements a Hades252 strategy for `Variable` as input values.
+/// Requires a reference to a `ConstraintSystem`.
+pub struct GadgetStrategy<'a> {
+    /// A reference to the constraint system used by the gadgets
+    pub cs: &'a mut StandardComposer<Curve>,
+}
+
+impl<'a> GadgetStrategy<'a> {
+    /// Constructs a new `GadgetStrategy` with the constraint system.
+    pub fn new(cs: &'a mut StandardComposer<Curve>) -> Self {
+        GadgetStrategy { cs }
+    }
+}
+
+impl<'a> Strategy<Variable> for GadgetStrategy<'a> {
+    fn quintic_s_box(&mut self, value: &mut Variable) {
+        let v_var = *value;
+        let v = self.cs.eval(value);
+
+        (0..4).for_each(|_| {
+            let o = self.cs.eval(value) * v;
+            let o_var = self.cs.add_input(o);
+
+            self.cs.mul_gate(
+                *value,
+                v_var,
+                o_var,
+                Scalar::one(),
+                -Scalar::one(),
+                Scalar::zero(),
+                Scalar::zero(),
+            );
+
+            *value = o_var;
+        });
+    }
+
+    fn mul_matrix(&mut self, values: &mut [Variable]) {
+        let zero = self.cs.add_input(Scalar::zero());
+        let mut product = [zero; WIDTH];
+
+        for j in 0..WIDTH {
+            for k in 0..WIDTH {
+                let a = self.cs.add_input(MDS_MATRIX[j][k]);
```

Comment:
I belive that if the matrix will be always the same and the constants also, we could use the MDS_MATRIX terms as circuit descriptors (selector poly coefs) in order to reduce the number of gates (when possible) imo.

Reply:
Maybe not needed now, but for sure we should go for something similar to what I suggested when we move this to the `plonk/master` branch.

Reply:
You mean using the matrix element as qm, ql or qr? That would reduce the number of inputs, but the number of gates would remain unchanged; I always have to perform the multiplication and addition.

Also, this is not going to be moved to `plonk/master`, afaik